Jetty 12 - Remove ISO-8859-1 fallback decoding during UTF-8 decoding

gregw opened the pull request:
This is a partial replacement for #9496, focused on simplification of the Utf8Appendable contract and removal of inconsistencies in it's use.

The `Utf8Appendable` class now does not throw unless in the `takeString` methods, which explicitly throw `CharacterCodingException` or use a passed in `Supplier<X extends Throwable>.

The `toString` method remains like the `StringBuilder` and `StringBuffer` `toString`, in that it just returns the string.  However, it now does not change state and so can be used for debug.  Equivalent to the previous `toPartialString()`

There is now a `complete()` method to handled incomplete sequences and a `toCompleteString()` convenience method. We should probably rename `takeString` to `takeCompleteString` to be consistent.

gregw committed:
Simplify Utf8Appendable signature

Allow for replacement characters without throwing

gregw committed:
Simplify Utf8Appendable signature

Allow for replacement characters without throwing

gregw commented:
@joakime if this approach proves workable, we need to extract/combine the goodness out of #9496, as I've just been looking at Utf8Appendable here, but have also removed the 8859-1 fallback, but without any rigour.

@lachlan-roberts there are websocket issues as I believe it was relying on fast fail behaviour of append throwing.   I have added some checks for `hasCodingErrors()` to achieve the same result, but obviously more work is required.

gregw commented:
@joakime what is the JVM decoding mechanism you alluded to last night?  We do have support for a `CharsetDecoder` in `DecoderStringBuilder`, but the API is very inefficient, requiring allocation of little byte arrays and little char sequences on every append.    Is there a better alternative?

lachlan-roberts committed:
fix failing tests in websocket-core-tests

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts commented:
@gregw I fixed all failing tests in websocket-core-tests, and there seemed to be no failures in websocket-jetty-tests ee10.
So I will wait for the CI build and see if there are any other failures.

joakime commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/UrlEncoded.java:
```diff
@@ -329,8 +329,7 @@ private static void decodeUtf8To(String query, int offset, int length, BiConsume
                 case '&':
-                    value = buffer.toReplacedString();
-                    buffer.reset();
+                    value = buffer.takeString(IllegalArgumentException::new);
```
Message-less Exceptions are horrible

gregw replied:
Agree. Specially for a generic exception type like IAE. CharacterCodingException doesn't take a message, but the name kind of says it.  Might need to ressurect NotUtf8Exception

joakime commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/UrlEncoded.java:
```diff
@@ -348,8 +347,7 @@ else if (value != null && value.length() > 0)
                     }
-                    key = buffer.toReplacedString();
-                    buffer.reset();
+                    key = buffer.takeString(IllegalArgumentException::new);
```
Ditto

joakime commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/Utf8Appendable.java:
```diff
@@ -89,19 +89,47 @@ public Utf8Appendable(Appendable appendable)
+    public void reset()
     {
         _state = UTF8_ACCEPT;
+        _codingErrors = false;
```
active codepoint needs reset too

joakime replied:
Done.

joakime commented on jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/CloseStatus.java:
```diff
@@ -168,21 +168,14 @@ public CloseStatus(ByteBuffer payload)
+                Utf8StringBuilder utf = new Utf8StringBuilder();
+                // if this throws, we know we have bad UTF8
+                utf.append(reasonBytes, 0, reasonBytes.length);
+                String reason = utf.takeString(() -> new BadPayloadException("Invalid UTF8 in CLOSE Reason", new CharacterCodingException()));
```
Huh?  why add the CharacterCodingException at all?

gregw replied:
It was for websocket tests that were checking cause. Happy to change tests and not have double exception, but wanted to start with minimal test changes

joakime commented on jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/WebSocketCoreSession.java:
```diff
@@ -319,7 +319,7 @@ public void processConnectionError(Throwable cause, Callback callback)
         if (cause instanceof CloseException)
             code = ((CloseException)cause).getStatusCode();
-        else if (cause instanceof Utf8Appendable.NotUtf8Exception)
+        else if (cause instanceof CharacterCodingException)
```
Just get rid of this if case entirely.
The new code throwing `BadPayloadException` is enough, as that class extends from `CloseException` already.

gregw replied:
Probably yes.

joakime commented on jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/WebSocketCoreSession.java:
```diff
@@ -355,7 +355,7 @@ public void processHandlerError(Throwable cause, Callback callback)
         else if (cause instanceof ClosedChannelException)
             code = CloseStatus.NO_CLOSE;
-        else if (cause instanceof Utf8Appendable.NotUtf8Exception)
+        else if (cause instanceof CharacterCodingException)
```
Just get rid of this if case entirely.
The new code throwing `BadPayloadException` is enough, as that class extends from `CloseException` already.

joakime commented on jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/internal/MessageHandler.java:
```diff
@@ -200,29 +200,29 @@ protected void onTextFrame(Frame frame, Callback callback)
             if (frame.isFin())
             {
-                onText(textBuffer.toString(), callback);
+                onText(textBuffer.takeString(CharacterCodingException::new), callback);
```
This should throw `BadPayloadException`, with a message.

joakime commented on jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/internal/MessageHandler.java:
```diff
@@ -200,29 +200,29 @@ protected void onTextFrame(Frame frame, Callback callback)
             }
         }
-        catch (Utf8Appendable.NotUtf8Exception e)
+        catch (CharacterCodingException e)
```
Remove this entire catch block (once you change to throwing a `BadPayloadException`)

joakime commented on jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/messages/PartialStringMessageSink.java:
```diff
@@ -42,12 +43,17 @@ public void accept(Frame frame, Callback callback)
             if (frame.isFin())
             {
-                methodHandle.invoke(out.toString(), true);
+                String complete = out.takeString(CharacterCodingException::new);
```
Throw `BadPayloadException` here.

joakime commented on jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/messages/PartialStringMessageSink.java:
```diff
@@ -42,12 +43,17 @@ public void accept(Frame frame, Callback callback)
-                methodHandle.invoke(out.takePartialString(), false);
+                String partial = out.toString();
+                if (out.hasCodingErrors())
+                    throw new CharacterCodingException();
```
Throw `BadPayloadException` here.

joakime commented on jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/messages/StringMessageSink.java:
```diff
@@ -50,8 +51,9 @@ public void accept(Frame frame, Callback callback)
-                methodHandle.invoke(out.toString());
-
+            {
+                methodHandle.invoke(out.takeString(CharacterCodingException::new));
```
Throw `BadPayloadException` here.

joakime commented on jetty-core/jetty-websocket/jetty-websocket-core-tests/src/test/java/org/eclipse/jetty/websocket/core/util/PartialStringMessageSinkTest.java:
```diff
@@ -94,8 +95,8 @@ public void testInvalidSingleFrameUtf8() throws Exception
-        RuntimeException error = assertThrows(RuntimeException.class, () -> callback.block(5, TimeUnit.SECONDS));
-        assertThat(error.getCause(), instanceOf(Utf8Appendable.NotUtf8Exception.class));
+        IOException error = assertThrows(IOException.class, () -> callback.block(5, TimeUnit.SECONDS));
+        assertThat(error.getCause(), instanceOf(CharacterCodingException.class));
```
`instanceOf(BadPayloadException.class)` here.

joakime commented on jetty-core/jetty-websocket/jetty-websocket-core-tests/src/test/java/org/eclipse/jetty/websocket/core/util/PartialStringMessageSinkTest.java:
```diff
@@ -114,8 +115,8 @@ public void testInvalidMultiFrameUtf8() throws Exception
-        RuntimeException error = assertThrows(RuntimeException.class, () -> continuationCallback.block(5, TimeUnit.SECONDS));
-        assertThat(error.getCause(), instanceOf(Utf8Appendable.NotUtf8Exception.class));
+        IOException error = assertThrows(IOException.class, () -> continuationCallback.block(5, TimeUnit.SECONDS));
+        assertThat(error.getCause(), instanceOf(CharacterCodingException.class));
```
`instanceOf(BadPayloadException.class)` here.

joakime commented on jetty-core/jetty-websocket/jetty-websocket-core-tests/src/test/java/org/eclipse/jetty/websocket/core/util/StringMessageSinkTest.java:
```diff
@@ -98,8 +99,8 @@ public void testInvalidSingleFrameUtf8() throws Exception
-        RuntimeException error = assertThrows(RuntimeException.class, () -> callback.block(5, TimeUnit.SECONDS));
-        assertThat(error.getCause(), instanceOf(Utf8Appendable.NotUtf8Exception.class));
+        IOException error = assertThrows(IOException.class, () -> callback.block(5, TimeUnit.SECONDS));
+        assertThat(error.getCause(), instanceOf(CharacterCodingException.class));
```
`instanceOf(BadPayloadException.class)` here.

joakime commented on jetty-core/jetty-websocket/jetty-websocket-core-tests/src/test/java/org/eclipse/jetty/websocket/core/util/StringMessageSinkTest.java:
```diff
@@ -118,8 +119,8 @@ public void testInvalidMultiFrameUtf8() throws Exception
-        RuntimeException error = assertThrows(RuntimeException.class, () -> continuationCallback.block(5, TimeUnit.SECONDS));
-        assertThat(error.getCause(), instanceOf(Utf8Appendable.NotUtf8Exception.class));
+        IOException error = assertThrows(IOException.class, () -> continuationCallback.block(5, TimeUnit.SECONDS));
+        assertThat(error.getCause(), instanceOf(CharacterCodingException.class));
```
`instanceOf(BadPayloadException.class)` here.

joakime commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/Utf8Appendable.java:
```diff
@@ -89,19 +89,47 @@ public Utf8Appendable(Appendable appendable)
-            int state = _state;
             _state = UTF8_ACCEPT;
-            throw new NotUtf8Exception("char appended in state " + state);
+            _codingErrors = true;
```
This method needs work.
This will now add a replacement character, then the caller will add it's own (bad) character.

gregw replied:
Isn't the replacement character for the incomplete sequence, then the caller adds it's own character, which can't be bad as it is already a character.

joakime replied:
The concern is that it could mean if you repeatedly pass the output back into Utf8Appendable the string grows, which is bad.
Imagine back/forth requests/responses like this.

gregw replied:
But we want it to grow. Somebody had called a method like append(char). We want to spend that char, but if we are incomplete then we append a replacement first.

joakime replied:
I think you misunderstand what I mean.

Take this example.

``` java
Utf8StringBuilder builder = new Utf8StringBuilder();
builder.append(stringFromRequest);
builder.complete();
String stringUsedInResponse = builder.toString();
```

Imagine that the client takes `stringUsedInResponse` and sends it back as `stringFromRequest` (don't get hung up on where in the request/response).
If that round trip makes the string grow every time then we have a bug.
The replacement character behavior should result in no growth of the string length when it is passed through the algorithm again.

gregw replied:
if you don't take or reset the builder, then it continues to grow.  Nothing new here???
So I think I do misunderstand what you mean.

joakime replied:
Imagine a string, supplied by an HTTP client, initially passed into Jetty as bad, then "sanitized" (with replacement characters).

We give that string back to the client (for whatever reason), in the "sanitized" form.

That same client gives us back that "sanitized" string in a different request.

That's the loop, a new Utf8StringBuilder each time, but the string on output is used again on a new Utf8StringBuilder (in append + toString)
That loop shouldn't result in a different String, or a longer String, as the replacement character should have sanitized it.
If each iteration through Utf8StringBuilder results in a different string, we have a bug.

gregw replied:
If we produce a decoded string, then validly encoder it and give it to the client, who then send it back to us, then this method will not add an extra replacement character because a) we don't add characters; b) it will now be well formed so a character will not be added when there is an incomplete code, this a replacement character will not be added.

joakime reviewed (changes requested):
First pass

joakime committed:
Adding more Utf8Appendable tests

joakime committed:
Actual messages in the URLEncoded exceptions

joakime committed:
Using CharacterCodingException properly for messages

joakime committed:
Correcting behavior with partial codepoints/sequences

joakime committed:
Fixing InvalidURIRule

joakime committed:
Address WebSocket behaviors with new Utf8Appendable

joakime committed:
Remove CharacterCodingException usage

gregw committed:
Simplify Utf8Appendable signature

minor reversions to minimise differences.

gregw commented:
@joakime thanks for the update.  However, I did revert a few minor changes, primarily to reduce needless differences with the original.  I believe I have kept all your behaviours.

gregw committed:
Simplify Utf8Appendable signature

Added violation to record bad UTF8

joakime commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/Utf8Appendable.java:
```diff
@@ -305,20 +299,20 @@ public void appendByte(byte b) throws IOException
                     }
                     _codep = 0;
-                    _state = current;
+                    _state = next;
```
BTW, this code is from http://bjoern.hoehrmann.de/utf-8/decoder/dfa/ (see javadoc)

The name "current" is the name used in the original code.
As is the decode() method.

joakime commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/UriCompliance.java:
```diff
@@ -69,7 +70,11 @@ public enum Violation implements ComplianceViolation
-        UTF16_ENCODINGS("https://www.w3.org/International/iri-edit/draft-duerst-iri.html#anchor29", "UTF16 encoding");
+        UTF16_ENCODINGS("https://www.w3.org/International/iri-edit/draft-duerst-iri.html#anchor29", "UTF16 encoding"),
+        /**
+         * Allow Bad UTF-8 encodings to be substituted by the replacement character.</code>.
```
This trailing `</code>` is breaking Javadoc

joakime reviewed (dismissed):
Javadoc portion of build is broken

joakime committed:
Fixing javadoc

lachlan-roberts commented on jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/messages/PartialStringMessageSink.java:
```diff
@@ -42,12 +43,17 @@ public void accept(Frame frame, Callback callback)
+                if (out.hasCodingErrors())
+                    throw new BadPayloadException("Invalid UTF-8");
+                out.partialReset();
+                methodHandle.invoke(partial, false);
```
This API seems weird that we don't have the same style for taking full and partial strings. 

Can't we add a method `takePartialString` in same style as `takeString` and avoid having these external `hasCodingError()` and `partialReset()` calls every time it is used.

```
out.takePartialString(() -> new BadPayloadException("Invalid UTF-8"));
```

gregw replied:
@lachlan-roberts it is a little complicated with the naming.
`takeString` should really be called `takeCompleteString`, but while that name makes sense for some character codes, it is not so much for things like iso-8859-1.  But then there are other multi-character codes served by `CharSequenceBuilder` interface, so perhaps doing that rename makes sense.

We could then re-introduce a `takeString` API that takes the partial string and matches to `toString` semantic.

@joakime @lachlan-roberts  thoughts?

lachlan-roberts replied:
I think its quite clear to have
- `takeString(Supplier<X>)`
- `takePartialString(Supplier<X>)`
- `toString()`

I'm okay with the rename of `takeString` -> `takeCompleteString`, but think it is unnecessary.
I think renaming `takePartialString` -> `takeString` is going to be very confusing.

joakime replied:
`takeString()` is a different API that `Utf8Appendable` implements via the `CharsetStringBuilder` interface, that name is already defined for other purposes.
It has to be a completed String.

joakime replied:
The fact that `toString()` is a partial String currently in this PR is actually tricky.  The overwhelming majority of uses of `toString()` are now bad and should be changed to `toCompleteString()`.

See commit f8b4884cfec8891d75ccd9aa642216cc1ab2ef3c here that fixes this bug.
(Submitted as a single commit to make it easier to revert)

gregw replied:
How many of those `toCompleteString` calls should really be `takeString`, except that they don't want a checked exception.... which can be achieved with the version that takes a supplier.

So leaning back towards a toString that gives debugging info.

Let me ponder some more on my next train.

joakime committed:
Fix Utf8AppendableTest

gregw commented:
If we reintroduce partial string in the naming, then it strange to also have toString be partial.

But toString can't be complete because it is a state change.

joakime committed:
Using toCompleteString() where appropriate

joakime commented:
> If we reintroduce partial string in the naming, then it strange to also have toString be partial.
> 
> But toString can't be complete because it is a state change.

Full circle back to what I had in the previous PR?

joakime committed:
Missing return statement

joakime committed:
Using correct exception type (per javadoc)

gregw commented:
Whilst playing with the API i realized that we no longer need Utf8StringBuffer.  So i got rid of it and collapsed Utf8Appendable into Utf8StringBuilder, which allowed me to get rid of all the needless IOException catches.

However, i note have a problem with the Null version, so I've got a bit of work to do. Might take me until tomorrow.

joakime commented:
> Whilst playing with the API i realized that we no longer need Utf8StringBuffer. So i got rid of it and collapsed Utf8Appendable into Utf8StringBuilder, which allowed me to get rid of all the needless IOException catches.

Yay!

joakime commented:
> However, i note have a problem with the Null version, so I've got a bit of work to do. Might take me until tomorrow.

The `NullAppender` is used by the `ValidationExtension`.

We we need need either of these anymore?
We do `takeString(() -> new BadPayloadException())` (and similar new Utf8Appendable APIs) in the websocket codebase later now.

gregw commented:
@lachlan-roberts what @joakime said?

gregw committed:
Simplify Utf8Appendable signature

Added some more to/take partial/complete methods for convenience.
Converted most usages of toString to the correct partial/complete method (might of missed some)
Made toString a debugging string.
Fixed some bugs and improved signature of CharsetStringBuilder (which assumed byte==char).

gregw committed:
Merge remote-tracking branch 'refs/remotes/origin/jetty-12.0.x-utf8Appendable-redux' into jetty-12.0.x-utf8Appendable-redux

Signed-off-by: gregw <gregw@webtide.com>

joakime commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/CharsetStringBuilder.java:
```diff
@@ -128,6 +176,42 @@ public void append(byte b)
+            }
+            catch (CharacterCodingException e)
+            {
+                // Should do replacement and remember issue.
```
This is the responsibility of the `CharsetDecoder` and how it's configured, we should honor its `onMalformedInput` configuration.

gregw replied:
Yep, didn't have time to work it out exactly. But yes, the decoder should do the replacement, not us.

gregw replied:
The behaviour was correct, so I have just updated the comment.

joakime commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/CharsetStringBuilder.java:
```diff
@@ -128,6 +176,42 @@ public void append(byte b)
+            }
+            catch (CharacterCodingException e)
+            {
+                // Should do replacement and remember issue.
```
This is the responsibility of the `CharsetDecoder` and how it's configured, we should honor its `onMalformedInput` configuration.

gregw replied:
The behaviour was correct, so I have just updated the comment.

gregw committed:
fixed decoderStringBuilder

gregw committed:
Merge branch 'jetty-12.0.x' into jetty-12.0.x-utf8Appendable-redux

gregw committed:
Improved CharsetStringBuilder consistency

Aligned behaviour for exceptions on various implementations.
Added javadoc to explain why the interface exists and how it works.

gregw committed:
Improved CharsetStringBuilder consistency

Aligned behaviour for exceptions on various implementations.
Added javadoc to explain why the interface exists and how it works.

gregw commented:
@joakime I have updated this PR so that when using `CharsetStringBuilder`, all the various implementations act the same (i.e. throw `CharacterCodingException` in the `takeString` method in a non-fast fail style).  I've also improved the javadoc to explain why the class exists.

The `Utf8StringBuilder` class should be used directly when we know it is UTF-8 and we want either fast fail or replacement behaviour.

gregw committed:
Improved CharsetStringBuilder consistency

Aligned behaviour for exceptions on various implementations.
Added javadoc to explain why the interface exists and how it works.

gregw commented:
@olamy This PR is the same - it is like CI is running the new test against the old code?

gregw committed:
Improved CharsetStringBuilder consistency

fixed tests

olamy commented:
> @olamy This PR is the same - it is like CI is running the new test against the old code?

not sure to understand the problem neither sure to understand which build you are referring to?
Looking at history of the build and commits before and after your comment

before your comment the build here https://jenkins.webtide.net/blue/organizations/jenkins/jetty.project/detail/PR-9507/16/tests is corresponding to this commit (https://github.com/eclipse/jetty.project/pull/9507/commits/850b65761885bfee8c86737ee923e41eb15ccb3d) which have test failures

before your comment the build here https://jenkins.webtide.net/blue/organizations/jenkins/jetty.project/detail/PR-9507/17/tests  is corresponding to this commit (https://github.com/eclipse/jetty.project/pull/9507/commits/bc9382122fe30c129bc12c9eaf625a64d9ce0b9b) which only change some javadoc

after your comment the build here https://jenkins.webtide.net/blue/organizations/jenkins/jetty.project/detail/PR-9507/18/pipeline which turn green is corresponding to this commit https://github.com/eclipse/jetty.project/pull/9507/commits/b307e6223bf493df378c3200ea69bfe64cd95bbb
last changes seems to fix some code and fix the build?

gregw commented:
@olamy yeah wrong PR sorry.  It is the session listener one that is causing me grief.

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-utf8Appendable-redux

gregw commented:
@joakime @lachlan-roberts OK green build finally.  Can I get reviews please.

joakime commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java:
```diff
@@ -1434,12 +1434,19 @@ else if (_path != null)
             }
         }
 
+        private RuntimeException onBadUtf8()
```
Why is this returning anything at all?
It always returns null below.

gregw replied:
Because it is being passed in as a `Supplier<X extends Throwable> onBadUtf8` to `URIUtil.canonicalPath`.

In this case we don't want to throw, so we supply null, but we remember that there was a violation

joakime commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/UriCompliance.java:
```diff
@@ -69,7 +70,11 @@ public enum Violation implements ComplianceViolation
          * Allow UTF-16 encoding eg <code>/foo%u2192bar</code>.
          */
-        UTF16_ENCODINGS("https://www.w3.org/International/iri-edit/draft-duerst-iri.html#anchor29", "UTF16 encoding");
+        UTF16_ENCODINGS("https://www.w3.org/International/iri-edit/draft-duerst-iri.html#anchor29", "UTF16 encoding"),
```
The hyphen matters.

s/UTF16/UTF-16/

joakime commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/UriCompliance.java:
```diff
@@ -69,7 +70,11 @@ public enum Violation implements ComplianceViolation
+        /**
+         * Allow Bad UTF-8 encodings to be substituted by the replacement character.
+         */
+        BAD_UTF8_ENCODING("https://datatracker.ietf.org/doc/html/rfc5987#section-3.2.1", "Bad UTF8 encoding");
```
The hyphen matters.

s/UTF8/UTF-8/

joakime commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/UrlEncoded.java:
```diff
@@ -873,7 +872,14 @@ else if (buffer != null)
             }
 
-            return buffer.toString();
+            try
+            {
+                return buffer.takeString();
+            }
+            catch (CharacterCodingException e)
+            {
+                throw new RuntimeException(e);
+            }
```
```suggestion
            return buffer.takeString(() -> new RuntimeException("Bad UTF-8"));
```

gregw replied:
Currently the API taking a `Supplier<X extends Throwable>` is only on the `Utf8StringBuilder` interface and not on  `CharsetStringBuilder`.  We don't have it there because we have some implementations (i.e. the JVM Decoder ones) that are hard wired to throw `CharacterCodingException`, so we cannot implement an arbitrary supplier.

joakime commented on jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/resource/FileSystemResourceTest.java:
```diff
@@ -183,6 +184,7 @@ public void testNonAbsoluteURI(WorkDir workDir) throws Exception
     }
 
     @Test
+    @Tag("flaky")
     public void testNotFileURI()
```
Is this flakiness filed as an issue?

gregw replied:
Ditto

joakime commented on jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/thread/QueuedThreadPoolTest.java:
```diff
@@ -317,6 +318,7 @@ public void testThreadPool() throws Exception
     }
 
     @Test
+    @Tag("flaky")
     public void testThreadPoolFailingJobs() throws Exception
```
Is this flakiness filed as an issue somewhere?

gregw replied:
No.  I assume that we will periodically search for flaky tags and work to fix them all.

joakime reviewed (commented):
Overall, I like it.
Just some nits and questions.

gregw committed:
Hyphenate

lachlan-roberts commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/CharsetStringBuilder.java:
```diff
@@ -40,13 +53,30 @@ default void append(byte[] b, int offset, int length)
 
+    /**
+     * @return The decoded built string.
+     * @throws RuntimeException if there is a {@link CharacterCodingException}.
```
Whats the point of this, why not just throw `CharacterCodingException` if there is a `CharacterCodingException`? And why doesn't `takeString` do the same?

Why is this called `toCompleteString` and the one below is not called `takeCompleteString`?

gregw replied:
This method is actually no longer needed in this interface.  so I have removed it.

There is a slight naming clash between the `CharsetStringBuilding` and `Utf8StringBuilder`.    The later already has a `takeCompleteString()` method that does not throw..... hmm let me see if I can resolve that.....

gregw replied:
OK, I renamed `takeString()` to `build()` as this is a builder after all.

gregw committed:
Removed unused methods. Some renaming

gregw committed:
Removed unused methods. Some renaming. Better NullAppendable
